src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs

```diff
@@ -2320,7 +2320,7 @@ public static void Refresh(Assembly assembly)
         public static Type ComObjectType
         {
             [PreserveDependency(".ctor", "System.ComponentModel.TypeDescriptor/TypeDescriptorComObject")]
-            [PreserveDependency(".ctor", "System.ComponentModel.TypeDescriptor/ComNativeDescriptorProxy")] // TODO: https://github.com/mono/linker/issues/801
+            [PreserveDependency(".ctor", "System.ComponentModel.TypeDescriptor/ComNativeDescriptorProxy")]
```

Comment:
We should replace these two `PreserveDependency` with `DynamicallyAccessedMembers`. The `DynamicallyAccessedMembers` annotation is preferred to `PreserveDependency` because it enrichens the metadata the linker sees. It helps prevent warnings showing up at the places where reflection actually happens.

E.g. if someone were to do `Activator.CreateInstance(TypeDescriptor.ComObjectType)`, the `DynamicallyAccessedMembers` annotation will ensure linker knows that the CreateInstance call is safe. `PreserveDependency` on the other hand wouldn't help, because when analyzing the CreateInstance call, linker doesn't look into the `ComObjectType` implementation and won't know that that's the thing we're keeping.

What should work here:
* Place `[return: DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.DefaultConstructor)]` on the getter. This will ensure that `TypeDescriptorComObject`'s constructor is kept because `typeof(TypeDescriptorComObject)` is the return value of this method.
* Annotate `TypeDescriptionProviderAttribute` (this is the connection from `TypeDescriptorComObject` to `ComNativeDescriptorProxy`): place `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.DefaultConstructor)]` on the `typeName` constructor parameter, `type` constructor parameter, and `TypeName` property. Doing this will ensure two things: `ComNativeDescriptorProxy`'s constructor is kept, and [this code](https://github.com/dotnet/runtime/blob/3bb5f14b31213c2d148505d7af1b4bc6b09a63e9/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs#L328-L340) stops generating warnings about potentially unsafe calls to `Activator.CreateInstance` (linker now sees that the call is safe)

Reply:
Oh, I see you already annotated TypeDescriptionProviderAttribute - the only thing that's missing is annotating the TypeName property.

Reply:
Thanks.

Do we need `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.DefaultConstructor)]` on the `TypeName` property, given it's the same type passed in the attribute's ctors (which already have the annotation, thus the type's default ctor was presumably already retained)?

Reply:
The annotation on TypeName is needed so that this code stops generating warnings: https://github.com/dotnet/runtime/blob/3bb5f14b31213c2d148505d7af1b4bc6b09a63e9/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs#L328-L340 (here we read that property, pass it to Type.GetType and Activator.CreateInstance). Without the annotation linker would say it's an unsafe operation.

It's not directly related to removing ILLinkTrim.xml, but might as well go all the way with this.